### DISCUSSION TITLE
Setting up blocky on VPS behind Nginx

### DISCUSSION THREAD

**Discussion by bytebone** [0xERR0R/blocky#649] [Q&A]
Hey there,

I've been trying to get blocky to work on my external server for a while now, but can't find helpful information anywhere, and don't know enough myself to fix it.

I've got blocky running on an external VPS, and from what I can tell it's running perfectly fine. Blocky is running inside a docker container. I've also got Nginx running, and set up a subdomain "dns.web.site" that forwards to the blocky container on port 853. This port is also openly exposed.
I only want to use DoT, which is why I don't have a configuration for port 53 (DoH). Both containers are in a shared network, so Nginx should be perfectly able to connect to blocky. 

I've tried adding the subdomain "dns.web.site" as a private DNS on my phone, but the connection fails. I've got no idea how to properly debug a DNS server. Certainly I'm missing some form of configuration in Nginx to allow DNS resolve traffic, but have no clue what to change. Maybe someone here knows what to do.

Thanks in advance.

**Comment by kwitsch** [comment]
I don't use nginx myself therefore this is only a guess but this [repo](https://github.com/TuxInvader/nginx-dns) should cover this use case.


Out of curiosity: what is the advantage routing DoT requests through nginx and not letting blocky handle these requests directly?

**Comment by omicron-b** [comment]
@RainerZufahl do you by chance use an Android device and a Let's Encrypt certificate?
There's [this issue](https://community.letsencrypt.org/t/android-devices-with-dot-configured-interaction-with-new-default-chain/161020) you may have encountered
The easy way to check if your DNS server is available from an Android device is by using this [third-party app](https://play.google.com/store/apps/details?id=com.frostnerd.smokescreen&gl=US)

**Comment by 0xERR0R** [comment]
Can you check with other tool (for example https://github.com/ogham/dog) if your DoT endpoint works properly, please? Android is not really good in DoT "debugging", because you have no logs.

I used DoT in the past and I had issues with LetsEncrypt and Android. The solution was to use "ISRG Root X1" as preferred chain (as mentioned by @omicron-b)

**Comment by bytebone** [comment]
> Out of curiosity: what is the advantage routing DoT requests through nginx and not letting blocky handle these requests directly?

I simply don't know how to do it otherwise. I guess I could use the server IP and enter it directly as DNS server in my router and such, but as I understood it, for DoT resolving there needs to be an attached domainname? And this domainname is managed through Nginx in my case.

> do you by chance use an Android device and a Let's Encrypt certificate? There's [this issue](https://community.letsencrypt.org/t/android-devices-with-dot-configured-interaction-with-new-default-chain/161020) you may have encountered The easy way to check if your DNS server is available from an Android device is by using this [third-party app](https://play.google.com/store/apps/details?id=com.frostnerd.smokescreen&gl=US)

Since I haven't supplied my own certificate, I assume blocky is using a selfgenerated Let's Encrypt certificate. So yeah, maybe what you linked applies here as well, but the issue is honestly too technical for me to grasp the problem/solution :( 

> Can you check with other tool (for example https://github.com/ogham/dog) if your DoT endpoint works properly, please? Android is not really good in DoT "debugging", because you have no logs.

I've installed and checked with dog:
```
> dog dns.web.site *

Status: NXDomain
Status: NXDomain
Status: Format Error
Status: Format Error
Status: NXDomain
Status: NXDomain
Status: NXDomain
Status: NXDomain
Status: NXDomain
Status: NXDomain
Status: NXDomain
Status: NXDomain
Status: NXDomain
  A dns.web.site.   26m19s   157.**.***.** (< external IP of my Server)
SOA                    1h00m00s A "a.root-servers.net." "nstld.verisign-grs.com." 2022090401 30m00s 15m00s 7d0h00m00s 1d0h00m00s
SOA                    1h00m00s A "a.root-servers.net." "nstld.verisign-grs.com." 2022090401 30m00s 15m00s 7d0h00m00s 1d0h00m00s
SOA                    1h00m00s A "a.root-servers.net." "nstld.verisign-grs.com." 2022090401 30m00s 15m00s 7d0h00m00s 1d0h00m00s
SOA                    1h00m00s A "a.root-servers.net." "nstld.verisign-grs.com." 2022090401 30m00s 15m00s 7d0h00m00s 1d0h00m00s
SOA                    1h00m00s A "a.root-servers.net." "nstld.verisign-grs.com." 2022090401 30m00s 15m00s 7d0h00m00s 1d0h00m00s
SOA                    1h00m00s A "a.root-servers.net." "nstld.verisign-grs.com." 2022090401 30m00s 15m00s 7d0h00m00s 1d0h00m00s
SOA                    1h00m00s A "a.root-servers.net." "nstld.verisign-grs.com." 2022090401 30m00s 15m00s 7d0h00m00s 1d0h00m00s
SOA                    1h00m00s A "a.root-servers.net." "nstld.verisign-grs.com." 2022090401 30m00s 15m00s 7d0h00m00s 1d0h00m00s
SOA                    1h00m00s A "a.root-servers.net." "nstld.verisign-grs.com." 2022090401 30m00s 15m00s 7d0h00m00s 1d0h00m00s

>dog -S dns.web.site * (< enforces DoT)

Error [network]: Connection refused (os error 61)
Error [network]: Connection refused (os error 61)
Error [network]: Connection refused (os error 61)
```
I'm got little clue what this information truly means, but since the DoT requests all didn't go through, I'd assume that DoT is not configured properly, and is in some way being blocked by Nginx (or anything else in the chain). 

> I used DoT in the past and I had issues with LetsEncrypt and Android. The solution was to use "ISRG Root X1" as preferred chain (as mentioned by @omicron-b)

Thanks for the advice. I'll try to see what this means and how I set it up. Will report back with findings. Thanks for all the help so far!

**Comment by 0xERR0R** [comment]
Blocky uses existing certificate and doesn't create any LE certificates. If you want to use Nginx for LetsEncrypt handling, your Nginx schould route traffic from port 853 to blocky. Since you want to use encryption from nginx, you can route traffic to port 53.

So basically:

Traffic outside ---(encrypted)---> Nginx (853) ---(plain)----> Blocky (53)

I don't know, how it can be configured in nginx. I hat this setup in the past with traeffik and it worked for me. The only one disadvantage was: you can't determine the real client-IP in blocky (blocky sees only IP from Nginx).

I think, currently your Nginx doesn't forward requests to blocky. Do you see something in blocky's log?


Btw. I used following self-made proxy for this use case: https://github.com/0xERR0R/dns-proxy. This tool acts as DoT server, creates LE certificate automatically (only DNS challenge) and routes traffic to Plain DNS or DoH resolver (like blocky). In DoH case, you can preserve the client ID from DoT (for client filtering and logging).
